tests/utils/test__select_column.py

```diff
@@ -0,0 +1,263 @@
+            "type3": pd.to_datetime(
+                [np.datetime64("2018-01-01"), datetime.datetime(2018, 1, 1)]
+            ),
+        }
+    )
+
+
+def test_type(df):
+    """Raise TypeError if `columns_to_select` is the wrong type."""
+    with pytest.raises(TypeError):
+        _select_columns(df, 2.5)
+    with pytest.raises(TypeError):
+        _select_columns(df, 1)
+    with pytest.raises(TypeError):
+        _select_columns(df, ("id", "M_start_date_1"))
+    with pytest.raises(TypeError):
+        _select_columns(df, [3, "id"])
+
+
+def test_strings(df):
+    """
+    Raise NameError if `columns_to_select` is a string
+    and does not exist in the dataframe's columns.
+    """
+    with pytest.raises(NameError):
+        _select_columns(df, "word")
+    with pytest.raises(NameError):
+        _select_columns(df, "*starter")
+
+
+def test_slice_dtypes(df):
+    """
+    Raise ValueError if `columns_to_select` is a slice instance
+    and either the start value or the stop value is not a string,
+    or the step value is not an integer.
+    """
+    with pytest.raises(ValueError):
+        _select_columns(df, slice(1, "M_end_date_2"))
+    with pytest.raises(ValueError):
+        _select_columns(df, slice("id", 2))
+    with pytest.raises(ValueError):
+        _select_columns(df, slice("id", "M_end_date_2", "3"))
+
+
+def test_slice_presence(df):
+    """
+    Raise ValueError if `columns_to_select` is a slice instance
+    and either the start value or the end value is not present
+    in the dataframe.
+    """
+    with pytest.raises(ValueError):
+        _select_columns(df, slice("Id", "M_start_date_1"))
+    with pytest.raises(ValueError):
+        _select_columns(df, slice("id", "M_end_date"))
+
+
+def test_callable(df):
+    """
+    Check that error is raised if `columns_to_select` is a
+    callable, and at lease one Series has a wrong data type
+    that makes the callable unapplicable.
+    """
+    with pytest.raises(TypeError):
+        _select_columns(df, object)
+
+
+def test_callable_returns_Series(df):
+    """
+    Check that error is raised if `columns_to_select` is a
+    callable, and returns a Series.
+    """
+    with pytest.raises(ValueError):
+        _select_columns(df, lambda x: x + 1)
+
+
+def test_callable_no_match(df):
+    """
+    Raise ValueError if `columns_to_select` is a callable, and
+    no boolean results are returned, when the callable is
+    applied to each series in the dataframe.
+    """
+    with pytest.raises(ValueError):
+        _select_columns(df, pd.api.types.is_float_dtype)
+
+    with pytest.raises(ValueError):
+        _select_columns(df, lambda x: "Date" in x.name)
+
+
+def test_regex_presence(df):
+    """
+    Raise ValueError if `columns_to_select` is a regex
+    and none of the column names match.
+    """
+    with pytest.raises(ValueError):
+        _select_columns(df, re.compile(r"^\d+"))
+
+
+class Test_Columns_not_List_Various_Inputs:
```

Comment:
Can I ask what your reasons were for adopting a class data structure here? Is it more concise written this way than as a suite of pytest functions? Would it be reasonable to rewrite them to test functions rather than test classes, to keep the testing style consistent? Having two styles (without reason) makes it more jarring for a newcomer to contribute than having a single consistent style.

Reply:
just me and shiny things :). Was going through the pytest docs and discovered them; looked nice to have similar tests in one class. Definitely reorganize them to functions instead.

Reply:
Oh, I see! Haha, I was wondering where that came from. :smile:

We can use pytest marks to help organize tests together. For example, in another project, I helped break out slow tests from fast tests, so that they could run in parallel. Please feel free to add marks as you see appropriate!